商品情報編集機能のブランチ

TakumiOtsuki-417 opened the pull request:
# What

Furimaにおいて一度出品された商品の情報を出品者が編集できる機能を実装。
また、各商品の編集画面など直接URL入力された場合の特定条件下のアクセス制御を追加。

# Why

商品の情報を変更できるようにすることで、出品者にとって最適な出品の条件を後から追求できるようにするため。
また、アクセス制御を追加することで出品者でない者による出品情報の編集や無効なURLへの遷移を防ぐため。

# Gyazoリンク

- ログイン状態の出品者は、商品情報編集ページに遷移できる動画
https://gyazo.com/d41670d6f40928f63b00b974354cda67
- 正しく情報を記入すると、商品の情報を編集できる動画
https://gyazo.com/f5c9e2902599c19ed9ed1ae36eaa98d3
- 入力に問題のある状態では商品の情報は編集できず、エラーメッセージが出力される動画
https://gyazo.com/7ca520cebe4ef1d476ff5f7294c3a3a1
- 何も編集せずに更新をしても画像無しの商品にならない動画
https://gyazo.com/acadd462f1c99897e75dc05e2d9900e7
- ログイン状態のユーザーであっても、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（「出品していない＝そもそも存在しない」商品の編集ページの場合）
https://gyazo.com/c0a26fde397d2e5f389b7e1cd5cf72e6
- ログイン状態のユーザーであっても、URLを直接入力して出品していない商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（「出品していない＝他者の出品した」商品の編集ページの場合）
https://gyazo.com/650e281f30470f0b54298937916a26dd
- ログアウト状態のユーザーは、URLを直接入力して商品情報編集ページへ遷移しようとすると、ログインページに遷移する動画
https://gyazo.com/664dd25b8eec52a37a5c7a09c72b8159
- 商品名やカテゴリーの情報など、すでに登録されている商品情報は商品情報編集画面を開いた時点で表示される動画（画像に関しては、表示されない状態で良い）
https://gyazo.com/843993911b570b70b796e81873956fcf

※「ログイン状態の出品者であっても、URLを直接入力して自身の売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画」は現時点で購入機能が未実装であるため用意なし。

TakumiOtsuki-417 committed:
商品編集ページ実装

・編集ページであるedit.html.erbを導入。コードを一部修正
・ルーティングとコントローラーへの記述を追加
・編集ページへのリンクを実装

TakumiOtsuki-417 committed:
アクセス制限追加

・出品されていない(＝存在しない)商品詳細ページなどへの遷移に対する制限
・他者の商品編集ページへの遷移に対する制限

TakumiOtsuki-417 committed:
rubocop実行

・不要なコメント削除
・rubocop実行

yutamasumoto commented on app/controllers/items_controller.rb:
```diff
@@ -17,9 +19,30 @@ def create
     @item = Item.find(params[:id])
   end
+  def edit
+    @item = Item.find(params[:id])
```
```
@item = Item.find(params[:id])
```
こちらshow,edit,updateにて重複した記載ございますのでメソッド化してbefore_actionで呼び出すようにしましょう。

yutamasumoto commented on app/controllers/items_controller.rb:
```diff
@@ -17,9 +19,30 @@ def create
+    end
+  end
+  def not_edit_by_other_user
+    unless current_user.id == Item.find(params[:id]).user_id
```
```
Item.find(params[:id])
```
こちらの記載ですが現状コントローラー内で`@item`という変数に同じ情報が入っているかと思います。
リファクタリングの観点からこちらの@itemを使って実装を行ったほうがより見やすいコードとなりますのでこちらの記載を変更してみましょう。

yutamasumoto reviewed (commented):
レビュー依頼ありがとうございます！
2件コメントしましたので、ご確認をお願いいたします！

TakumiOtsuki-417 committed:
指摘点修正１

・show/edit/updateアクションで重複した記述をリファクタリング

TakumiOtsuki-417 committed:
指摘点修正２

・プライベートメソッド「find_item」追加により実質重複した記述を修正または削除

TakumiOtsuki-417 commented:
コードレビュー有難うございました！
コメントいただきました２点に関した修正を行ったので、お手数ですが再びご確認お願いいたします。

- プライベートメソッド「find_item」に重複した記述をまとめる（items_controller.rb）
- プライベートメソッド作成により不要になったと思われる重複記述を修正・削除（items_controller.rb）

yupimaru commented on app/controllers/items_controller.rb:
```diff
@@ -15,11 +17,33 @@ def create
   end
 
   private
+  def find_item
+    @item = Item.find(params[:id])
+  end
```
```ruby
  def find_item
    @item = Item.find(params[:id])
  end
```

priveteメソッドとして、切り分けることができましたね！
このメソッドを、before_actionで呼び出し、必要なアクション(show,edit,update)の場合のみ
適用されるように追記しましょう！

また、before_actionは記述した順番で処理が実行されるため、
`authenticate_user!`の次に実行されるように記述しておきましょう。

yupimaru commented on app/controllers/items_controller.rb:
```diff
@@ -15,11 +17,33 @@ def create
+    end
+  end
+  def not_edit_by_other_user
+    unless current_user.id == find_item.user_id
```
```ruby
current_user.id == find_item.user_id
```

こちらで使用している`find_item`ですが、
find_item自体をbefore_actionで呼び出すことにより、
`@item`と書き換えることが可能かと存じます！

yupimaru commented on app/controllers/items_controller.rb:
```diff
@@ -15,11 +17,33 @@ def create
   end
   def show
-    @item = Item.find(params[:id])
+    find_item
```
```ruby
find_item
```

こちらの記述でございますが、`find_item`自身をbefore_actionで呼び出すことにより、
削除することができそうです！

yupimaru reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
とても良くなっていますね！

追加で3点コメントいたしましたので、ご確認お願いします！

TakumiOtsuki-417 committed:
指摘点修正3

・「find_item」メソッドをbefore_actionとして、「not_edit_by_other_user」メソッド以前に呼び出し
・上記変更に関連して一部の記述を変更・削除

TakumiOtsuki-417 commented:
更なるレビュー、有難うございました。
いただいたコメントに関連した修正を行いましたので、お手数ですが確認いただきたく思います。

- before_actionとして「find_item」メソッドを呼び出し
- 上記変更により発生する不適箇所を変更・削除

strongmasahiro commented on app/views/items/edit.html.erb:
```diff
@@ -0,0 +1,149 @@
+        <p>
+          クリックしてファイルをアップロード
+        </p>
+        <%= f.file_field :hoge, id:"item-image" %>
```
`<%= f.file_field :hoge, id:"item-image" %>`
こちらのhogeの部分を修正しましょう！

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます！
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いします！

TakumiOtsuki-417 committed:
指摘点修正４

・ビューファイル修正

TakumiOtsuki-417 commented:
コードレビュー、有難うございます。見落とし失礼いたしました。
修正いたしましたので、再度ご確認いただけますと幸いです。

- edit.html.erb内に残っていた「hoge」部分を修正。該当箇所の編集が適用されること確認。

strongmasahiro reviewed (commented):
レビュー依頼ありがとうございます！
LGTMとさせていただきます！実装お疲れ様でした！
引き続き実装頑張っていきましょう！
